### PR TITLE
lint: Add commit/PR body linter for epic and issue refs

### LINKED ISSUE #77376: linter: Add PR check to encourage including issue or epic refs in the PR/commit messages
**Describe the solution you'd like**

This new linter should be a program run by a GitHub action on every PR for these events: create, edit, reopen, synchronize.

What it will do:

For all PRs containing commits with [Release note texts](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/186548364/Release+notes), other than bug fixes, Engineers can help establish the link to Jira epics in various ways:

- If the PR relates to only one Jira epic:
    - Put the GH issue or Jira epic reference either in the PR description or in each relevant commit.
    - If the reference is in the PR description, put “Epic: None” in each commit that doesn’t relate to the epic. This is particularly important if the commit contains a release note.
- If the PR relates to multiple Jira epics:
    - Put the GH issue or Jira epic reference in each relevant commit.
    - If a single commit relates to multiple Jira epics, put all relevant GH issue or Jira epic references in that commit.
    - If all commits relate to all the Jira epics, put the GH issue or Jira epic references in the PR description or in the commit messages.
    - Put “Epic: None” in each commit with a release note that doesn’t relate to any epic.
- If the PR relates to no Jira epics:
    - Preferably, put “Epic: None” in the PR description or in each commit containing a release note.

Given Engineering’s practice to [link GitHub issues to epics](https://cockroachlabs.atlassian.net/wiki/spaces/JIRA/pages/1925283849/Epic+Linking), referencing such an issue—as opposed to the epic it links to—is preferred in cases where both a Jira epic and GitHub issue exist. This makes information in the issue easily accessible to human readers while offering an epic link that is available via the issue to automation scripts.

The “Epic: None” option accounts for the fact that a significant amount of product changes occur spontaneously (serendipitously) as a by-product of investigations on main roadmap items. When this happens, Engineers take the initiative to make user-facing changes without prior processing by a PM, and thus without prior existence of a GH or Jira issue. We do not wish to reduce the ability of Engineers to innovate and improve our product in this way. 

This solution will be encouraged, but not required, by a new optional blathers check on PRs. It will fail and give feedback on how to adjust the PR for the following cases:

- When no GH issue reference, Jira epic reference, or “Epic: None” exists in the PR description or commit messages.
- When multiple Jira epic references are in the PR description but not in individual commit messages.
- When multiple Jira epic references are in a commit message (give a warning that it’s not a common case and to check that they intend it).
- When backport PRs backport multiple source PRs and one or more of the source PRs has one of the above cases.

**Describe alternatives you've considered**

This linter can't run in TeamCity because it needs to run on PR edit and reopen events. We considered using Blathers but decided to keep this in the cockroach repo.

Jira issue: CRDB-13558

Epic DEVINF-261

### PR DESCRIPTION
**Pull request by jlinder**
Before: Commits and PRs would inconsistently contain references to issues.
Sometimes the author would forget to add it. Some authors might not
think to add them. And there were no epic references provided.

Why: Adding issue and epic references to PRs and commits provides
traceability and context for the revenue teams, product, management and
engineers.

Now: A new GitHub Action workflow checks for references to issues the PR
closes or informs and epics the PR is part of. It looks for references
in the PR body and in commit messages and, if it doesn't find what it
expects, it fails the check.

Informs #77376

Release note: None

Release justification: Adding linter for better epic tracking

### CODE REVIEW

**Comment by jlinder** [comment]
@celiala @rail Here's a draft of the epic and issue reference linter. The structure of it is in good shape and the GitHub Actions workflow works. Lots of TODOs still. You can see an example of the action running in https://github.com/jlinder/cockroach/pull/2.

File: pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+	if err != nil {
+		// nolint:errwrap
+		return fmt.Errorf("error getting pull requests from GitHub: %v", err)
+	}
+	prBody := pr.GetBody()
+
+	// Skip checking multi-PR backports for now. Verifying them is complex and the source PRs
+	// have already been checked.
+	if isMultiPrBackport(prBody) {
+		// TODO(jamesl): handle backport PRs containing multiple source PRs differently
```
Comment (jlinder) [inline pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go:267]: I'm planning to not do this in the initial PR and come back to this later.

**Comment by jlinder** [comment]
@celiala @rail This PR is ready for an initial review pass. There are a couple small TODOs I have left to do.

**Review by rail — CHANGES_REQUESTED** [review]
> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77654)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @celiala and @jlinder)

___
*[`.github/workflows/pr-epic-issue-ref-lint.yml` line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N-VSw9c3y8raru1ooe_:-N-VSw9c3y8raru1ooea:b-ukkwqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/.github/workflows/pr-epic-issue-ref-lint.yml#L8)):*
> ```YAML
> 
> jobs:
>   test:
> ```

A nit. Maybe instead of "test" use some more meaningful name? lint-commit-message maybe?
___
*[`.github/workflows/pr-epic-issue-ref-lint.yml` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N-VTVK0Dpx_aGz49Upe:-N-VTVK0Dpx_aGz49Upf:b-r60goq) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/.github/workflows/pr-epic-issue-ref-lint.yml#L18)):*
> ```YAML
>         uses: actions/cache@v2
>         env:
>           version: "22.1"
> ```

What does "22.1" mean here? Assuming the current CRDB version, would it mean that we need to update the value? Maybe just get rid of it?
___
*[`.github/workflows/pr-epic-issue-ref-lint.yml` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N-VU7Ln6M28kE5M4OYS:-N-VU7Ln6M28kE5M4OYT:bjm3gc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/.github/workflows/pr-epic-issue-ref-lint.yml#L26)):*
> ```YAML
>       - name: Install bazelisk
>         run: |
>           curl -LO "https://github.com/bazelbuild/bazelisk/releases/download/v1.10.1/bazelisk-linux-amd64"
> ```

Can you add a sha256  integrity check step please.
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g1r6n7g8tdivKQNcf:-N0g1r6n7g8tdivKQNcg:bsw3tll) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L97)):*
> ```Go
> ) ([]commitInfo, error) {
> 	commits, _, err := ghClient.PullRequests.ListCommits(
> 		ctx, params.Org, params.Repo, params.PrNumber, &github.ListOptions{PerPage: 100},
> ```

Do we care about PRs with more than 100 commits? :D
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g2NtgDa6xeEHUvU9P:-N0g2NtgDa6xeEHUvU9Q:bciltlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L131)):*
> ```Go
> func checkForMissingRefs(
> 	prInfo prBodyInfo, commitInfos []commitInfo, params *parameters,
> ) (passedCheck bool) {
> ```

`passedCheck` is unused and may be confusing. Can we remove it so we fail if there is a bare return?



_Code quote:_
```Go
passedCheck
```
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g38450iFgvvTZXJpv:-N0g38450iFgvvTZXJpw:b-1w8a3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L169)):*
> ```Go
> 	printHeader := func() {
> 		if !passedCheck {
> 			return
> ```

Can we avoid using bare returns in this function? Looks like you can `return false` here.
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 231 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g3dk37RwpyaGvPyh7:-N0g3dk37RwpyaGvPyh8:btq2uaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L231)):*
> ```Go
> // Returns:
> // - True if the check passed. False if it failed.
> func checkMultipleEpicsInCommits(commitInfos []commitInfo, params *parameters) (passedCheck bool) {
> ```

Can we remove `passedCheck` from the function definition? it's redefined in the next line, so better to leave a foot gun ;)
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g4JK80gwATrxKcede:-N0g4JK80gwATrxKcedf:b-6thnze) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L294)):*
> ```Go
> 	failedATest := !checkForMissingRefs(prInfo, commitInfos, params)
> 	failedATest = !checkPrEpicsUsedInCommits(prInfo, commitInfos, params) || failedATest
> 	failedATest = !checkMultipleEpicsInCommits(commitInfos, params) || failedATest
> ```

can you simplify the representation of this to something like 
```go
failedTest := !(func1(...) && func2(...) ...)
```
or
```go
allTestsPassed := func1(...) && func2(...) && ...
```

___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g-xvJ7yViS39zMhpK:-N0g-xvJ7yViS39zMhpL:b-ugll1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L27)):*
> ```Go
> 	Use:   "lint-epic-issue-refs PR_NUMBER",
> 	Short: "Check the body and commit messages of the indicated PR for epic and issue refs",
> 	Args: func(cmd *cobra.Command, args []string) error {
> ```

You can replace it with
```go
Args: cobra.ExactArgs(1),
```
It won't do the integer parsing, but the parsing function will be simpler.
___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0fz-_n9P7UXvlWYRxy:-N0fz-_n9P7UXvlWYRxz:biuol2) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L34)):*
> ```Go
> 		params := defaultEnvParameters()
> 		prNumber, err := parseArgs(args)
> 		if err != nil {
> ```

Not a biggie, you can use `RunE` and return the error and handle the error in `main` instead of `os.Exit()` every time.
___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g0yU39HrVJZ024u3x:-N0g0yU39HrVJZ024u3y:b-tr2vl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L63)):*
> ```Go
> 	prNumber, err := strconv.Atoi(args[0])
> 	if err != nil {
> 		// nolint:errwrap
> ```

You can get rid of this if you `s/v/w/`
```go
return -1, fmt.Errorf("invalid PR number: %w", err)
```
___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g0mN0-C4HC13oJ9pd:-N0g0mN0-C4HC13oJ9pe:bsmati2) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L64)):*
> ```Go
> 	if err != nil {
> 		// nolint:errwrap
> 		return -1, fmt.Errorf("invalid PR number: %v", err)
> ```

Maybe stick with Go's default value "0" for integers?

**Review by jlinder — COMMENTED** [review]
Thanks for the feedback @rail .

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-:-N1RgPle--ed6sytB9Xj:b-ecjtfn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @celiala and @rail)

___
*[`.github/workflows/pr-epic-issue-ref-lint.yml` line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N-VSw9c3y8raru1ooe_:-N1RLaMY2dcSMZ6kHiWP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/.github/workflows/pr-epic-issue-ref-lint.yml#L8)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

A nit. Maybe instead of "test" use some more meaningful name? lint-commit-message maybe?
</blockquote></details>

Done.
___
*[`.github/workflows/pr-epic-issue-ref-lint.yml` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N-VTVK0Dpx_aGz49Upe:-N1RjsMIAhPNwJDm0GhR:bqqopuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/.github/workflows/pr-epic-issue-ref-lint.yml#L18)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

What does "22.1" mean here? Assuming the current CRDB version, would it mean that we need to update the value? Maybe just get rid of it?
</blockquote></details>

Thanks for asking that. We can get rid of it. I remember following an example that had a version in it and I see it's not necessary here.
___
*[`.github/workflows/pr-epic-issue-ref-lint.yml` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N-VU7Ln6M28kE5M4OYS:-N1RggXl8mEaEp_F-_cj:byvyago) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/.github/workflows/pr-epic-issue-ref-lint.yml#L26)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Can you add a sha256  integrity check step please.
</blockquote></details>

Yes. Done.
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g1r6n7g8tdivKQNcf:-N1RVLS8CfAMxLN8hnzk:bf2o9c0) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L97)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Do we care about PRs with more than 100 commits? :D
</blockquote></details>

Hehe. The only time I've seen around or more than 100 commits in a PR was for a bulk backporting PR. Or Maybe I just haven't seen some folks make such a large PR. 

Added paging. This API endpoint has a limit of 250 commits it would return. Thus the limit of only querying for three pages 
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g2NtgDa6xeEHUvU9P:-N1RNkQJ2lATtkzbJpjz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L131)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

`passedCheck` is unused and may be confusing. Can we remove it so we fail if there is a bare return?
</blockquote></details>

Done.
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g38450iFgvvTZXJpv:-N1ROHL95LdJLiRIau_I:b-6y179y) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L169)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Can we avoid using bare returns in this function? Looks like you can `return false` here.
</blockquote></details>

This return is in a closer function that doesn't have a return parameter so it can't return a value.
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 231 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g3dk37RwpyaGvPyh7:-N1RPfr3EjWqr7577W_9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L231)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Can we remove `passedCheck` from the function definition? it's redefined in the next line, so better to leave a foot gun ;)
</blockquote></details>

Done.
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go` line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g4JK80gwATrxKcede:-N1RQ9HB0ozF5m1RwhLc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go#L294)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

can you simplify the representation of this to something like 
```go
failedTest := !(func1(...) && func2(...) ...)
```
or
```go
allTestsPassed := func1(...) && func2(...) && ...
```

</blockquote></details>

Done.
___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g-xvJ7yViS39zMhpK:-N1RYySv3D7qUlffz7R9:b-s6xupk) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L27)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

You can replace it with
```go
Args: cobra.ExactArgs(1),
```
It won't do the integer parsing, but the parsing function will be simpler.
</blockquote></details>

Oh, that's nice. Done.
___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0fz-_n9P7UXvlWYRxy:-N1RaTgNDKlCecWz6zVF:b-afvmjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L34)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Not a biggie, you can use `RunE` and return the error and handle the error in `main` instead of `os.Exit()` every time.
</blockquote></details>

I did this originally and when an error happened, it would also print out the usage information for the command. I didn't want that for when the Run function produces an error. So I manually handled the errors in the Run function. Now I see there is a `SilenceUsage` option but it also silences usage for the ExactArgs(1) check. If there's a way to SilenceUsage only for errors from RunE, then I'm down for making the change (I didn't see an obvious way on a quick look through the docs).
___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g0yU39HrVJZ024u3x:-N1RZsud63-jzXLsHI3e:bkkra1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L63)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

You can get rid of this if you `s/v/w/`
```go
return -1, fmt.Errorf("invalid PR number: %w", err)
```
</blockquote></details>

:thumbsup: Thanks.
___
*[`pkg/cmd/lint-epic-issue-refs/main.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-N0g0mN0-C4HC13oJ9pd:-N1R_BzS57K0BPF54j0x:bh7t6wr) ([raw file](https://github.com/cockroachdb/cockroach/blob/abaf8f41b95b4723ab87fdc7cffe28594cc5213c/pkg/cmd/lint-epic-issue-refs/main.go#L64)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Maybe stick with Go's default value "0" for integers?
</blockquote></details>

That makes sense. Done.

File: pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+	"reflect"
+	"regexp"
+
+	"github.com/google/go-github/github"
+	"golang.org/x/oauth2"
+)
+
+var (
+	fixIssueRefRE    = regexp.MustCompile(`(?im)(?:close[sd]?|fix(?:e[sd])?|resolve[sd]?):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+))[,.;]?(?:[ \t]+|$))+`)
+	informIssueRefRE = regexp.MustCompile(`(?im)(?:see also|informs):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+))[,.;]?(?:[ \t]+|$))+`)
```
Comment (jlinder) [inline pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go:25]: I figured out that we also want to support issue refs for Jira issues. The easiest way to do that is:

- Update these two regexes to also match Jira issue keys
- Add tests that include Jira issue refs
- Ensure Jira issue refs are handled correctly in the code that uses these regexes and the code that consumes the found issue refs.

I think the updated regexes would look like this:

```suggestion
	fixIssueRefRE    = regexp.MustCompile(`(?im)(?:close[sd]?|fix(?:e[sd])?|resolve[sd]?):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+)|([A-Z]+-[0-9]+))[,.;]?(?:[ \t]+|$))+`)
	informIssueRefRE = regexp.MustCompile(`(?im)(?:see also|informs):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+)|([A-Z]+-[0-9]+))[,.;]?(?:[ \t]+|$))+`)
```

**Comment by nickvigilante** [comment]
@nickvigilante Fixes https://github.com/cockroachdb/cockroach/issues/77376

**Review by rail — APPROVED** [review]
> Reviewed 3 of 3 files at r2, 2 of 2 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77654)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @celiala)

**Review by jlinder — COMMENTED** [review]
Looking good. A few comments.

This can be squashed down into the single commit for the PR. With Reviewable (and for the most part with GitHub reviews) it's not necessary to make separate commits for changes like this. The difference can easily be seen when the same single commit is force pushed to.

> Reviewed 2 of 8 files at r1, 2 of 2 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-:-NAHu8FJ7gUpr_yXuBot:bun540h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @celiala)

___
*[`.github/workflows/pr-epic-issue-ref-lint.yml` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-NAHqip18AE7nJB0XQWZ:-NAHqip18AE7nJB0XQW_:b-hsk9uh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e8ce6e63b23a4f1bb3a010ddc32a8784f1579a9/.github/workflows/pr-epic-issue-ref-lint.yml#L1)):*
> ```YAML
> name: Epic and Issue Reference Linter for PR Body or Commit Messages
> ```

Maybe something like this for a slightly shorter name?

```suggestion
name: Epic + Issue Ref Linter for PR Body or Commit Messages
```
___
*[`pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs_test.go` line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77654#-NAHs2hRAfK0jiw25Unp:-NAHs2hRAfK0jiw25Unq:b-1jrluc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e8ce6e63b23a4f1bb3a010ddc32a8784f1579a9/pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs_test.go#L128)):*
> ```Go
> 
> Fixed:  CRDB-12345 example/repo#941
> informs: doc-1234, crdb-12345
> ```

Could one of the `CRDB-12345` refs be changed to a different number? While they are technically different because of capitalization, they can be confused by someone reading the code if they are only looking at the number portion (I missed they were different for that reason on the first read).

**Comment by nickvigilante** [comment]
I created https://github.com/cockroachdb/cockroach/pull/87107 in my own repo to continue development.

**Comment by celiala** [comment]
Removing myself as a reviewer, since it looks like we're going to eventually close this out in favor of https://github.com/cockroachdb/cockroach/pull/87128

**Pull request closed without merging**